Sources/NIO/BaseSocketChannel.swift

```diff
@@ -1095,7 +1095,9 @@ class BaseSocketChannel<SocketType: BaseSocketProtocol>: SelectableChannel, Chan
 
             return readStreamState
         }
+        #if !SWIFTNIO_USE_IO_URING
```

Comment:
So I've lost track of where we are with this: @hassila do you feel confident this is not masking a bug?

Reply:
Yes, I think that the root problem here is that the io_uring backend does not have the implicit synchronisation between modifications to the poll mask and the actual returned events on the completion queue that kqueue and epoll has. For kqueue and epoll, there is an implicit synchronisation point such that after a modification of the poll mask has been issued, the next call to reap events will be sure to not include events which does not match the new poll mask. Specifically for this assert, it means that we will be guaranteed to never receive a POLLIN notification unless there are bytes available to read.

For a fully asynchronous backend like io_uring, there are no such implicit synchronisation point, so after we have submitted the asynchronous event to change the poll mask, we may still reap pending asynchronous replies for the old poll mask, and thus receive a POLLIN even though we have modified the mask visavi the kernel.

The only way to avoid that race, would be to use heavy handed synchronisation primitives like IOSQE_IO_DRAIN (basically flushing **all** pending requests and wait for a fake event result to sync up) - see some discussions at (LWN for fsync)[https://lwn.net/Articles/785690/]. It would be awful from a performance POV to get that guarantee, so would definitely not prefer to assert() this (as the assert simply is not valid for the full asynchronous backend case without implicit synchronisation points).

Reply:
So long as we don't think this is covering up a bug I'm fine with that. Can we add a code comment with this rationale though?

Reply:
Sure.